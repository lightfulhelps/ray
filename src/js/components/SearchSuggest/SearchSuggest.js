// @flow
import * as React from 'react';
import classNames from 'classnames';
import { DropdownMenu, DropdownItem, Icon } from '../../';

type Props = {
  className?: string,
  isLoading?: boolean,
  isOpen: boolean,
  limit?: number,
  onClear?: () => void,
  onClick?: () => void,
  onRemove?: number => void,
  onSelect: (string, number) => void,
  options: string[],
  search: string,
  title?: string,
};

export const findMatches = (option: string, search: string): boolean =>
  option.toLowerCase().includes(search.toLowerCase());

export const highlightMatches = (option: string, search: string): string => {
  if (!search) return option;

  const regex = new RegExp(`(${search})`, 'gi');
  const ret = option.replace(regex, `<strong class="text-gray-900">$&</strong>`);

  return ret;
};

const SearchSuggest = ({
  className,
  isLoading,
  isOpen,
  limit,
  onClear,
  onClick,
  onRemove,
  onSelect,
  options = [],
  search = '',
  title,
  ...other
}: Props) => {
  const classes = classNames(className, 'search-suggest');
  const count = limit && limit > 0 ? limit : 10;
  const output = options.filter(option => findMatches(option, search)).splice(0, count);

  if (!search.length && !output.length) return null;

  return (
    <DropdownMenu
      {...other}
      className={classes}
      data-test-id="search-suggest"
      isOpen={isOpen}
      onClick={onClick}
    >
      {title && (
        <DropdownItem
          className="d-flex justify-content-between align-items-center"
          data-test-id="search-suggest-header"
          isHeader
        >
          {title}{' '}
          {!isLoading &&
            onClear && (
              <div
                className="ml-1 text-danger cursor-pointer text-capitalize"
                data-test-id="search-suggest-clear"
                onClick={onClear}
              >
                Clear All
              </div>
            )}
        </DropdownItem>
      )}
      {isLoading && (
        <DropdownItem data-test-id="search-suggest-loading" isDisabled>
          Loading...
        </DropdownItem>
      )}
      {!isLoading &&
        search.length > 0 &&
        output.length === 0 && (
          <DropdownItem data-test-id="search-suggest-empty">
            No results for &ldquo;
            {search}
            &rdquo;
          </DropdownItem>
        )}
      {!isLoading &&
        output.length > 0 &&
        output.map((option, i) => (
          <DropdownItem
            className="d-flex justify-content-between align-items-center text-body"
            data-test-id="search-suggest-item"
            key={i}
            onClick={() => onSelect(option, i)}
          >
            <div dangerouslySetInnerHTML={{ __html: highlightMatches(option, search) }} />
            {onRemove && (
              <Icon
                className="ml-1"
                data-test-id="search-suggest-remove"
                name="close"
                onClick={e => {
                  e.stopPropagation();
                  onRemove(i);
                }}
              />
            )}
          </DropdownItem>
        ))}
    </DropdownMenu>
  );
};

export default SearchSuggest;
