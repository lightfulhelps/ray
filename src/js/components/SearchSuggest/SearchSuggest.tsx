import * as React from 'react';
import classNames from 'classnames';
import { DropdownMenu, DropdownItem, Icon } from '../..';

type ExcludeType = string | RegExp;

type Props = {
  className?: string;
  exclude: ExcludeType;
  isLoading?: boolean;
  isOpen: boolean;
  limit?: number;
  onClear?: () => void;
  onClick?: () => void;
  onRemove?: (arg0: number) => void;
  onSelect: (arg0: string, arg1: number) => void;
  options: string[];
  search: string;
  title?: string;
};

export const findMatches = (option: string, search: string, exclude: ExcludeType): boolean => {
  const normalize = (str: string) => str.replace(exclude, '').toLowerCase();

  return normalize(option).includes(normalize(search));
};

export const highlightMatches = (option: string, search: string, exclude: ExcludeType): string => {
  if (!search) return option;

  const regex = new RegExp(`(${search.replace(exclude, '')})`, 'gi');
  const ret = option.replace(regex, `<strong class="text-gray-900">$&</strong>`);

  return ret;
};

const SearchSuggest: React.FC<Props> = ({
  className,
  exclude,
  isLoading,
  isOpen,
  limit,
  onClear,
  onClick,
  onRemove,
  onSelect,
  options = [],
  search = '',
  title,
  ...other
}: Props) => {
  const classes = classNames(className, 'search-suggest');
  const count = limit && limit > 0 ? limit : 10;
  const filteredOptions = options.filter(option => findMatches(option, search, exclude));

  if (!isLoading && !options.length) return null;
  if (search.length > 0 && !filteredOptions.length) return null;

  return (
    <DropdownMenu
      {...other}
      className={classes}
      data-test-id="search-suggest"
      isOpen={isOpen}
      onClick={onClick}
    >
      {title && (
        <DropdownItem
          className="d-flex justify-content-between align-items-center"
          data-test-id="search-suggest-header"
          isHeader
        >
          {title}{' '}
          {!isLoading && onClear && (
            <div
              className="ml-1 text-danger cursor-pointer text-capitalize"
              data-test-id="search-suggest-clear"
              onClick={onClear}
            >
              Clear All
            </div>
          )}
        </DropdownItem>
      )}
      {isLoading && (
        <DropdownItem data-test-id="search-suggest-loading" isDisabled>
          Loading...
        </DropdownItem>
      )}
      {!isLoading &&
        filteredOptions.slice(0, count).map((option, i) => (
          <DropdownItem
            className="d-flex justify-content-between align-items-center text-body"
            data-test-id="search-suggest-item"
            key={i}
            onClick={() => onSelect(option, i)}
          >
            <div dangerouslySetInnerHTML={{ __html: highlightMatches(option, search, exclude) }} />
            {onRemove && (
              <Icon
                className="ml-1"
                data-test-id="search-suggest-remove"
                name="close"
                onClick={(e: React.MouseEvent) => {
                  e.stopPropagation();
                  onRemove(i);
                }}
              />
            )}
          </DropdownItem>
        ))}
    </DropdownMenu>
  );
};

export default SearchSuggest;
