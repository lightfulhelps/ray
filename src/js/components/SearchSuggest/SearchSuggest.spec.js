import React from 'react';
import { shallow } from 'enzyme';
import merge from 'lodash/merge';
import SearchSuggest, { findMatches, highlightMatches } from './SearchSuggest';

const setup = (overrides = {}) => {
  const props = merge(
    {
      isLoading: false,
      onClear: jest.fn(),
      onClick: jest.fn(),
      onRemove: jest.fn(),
      onSelect: jest.fn(),
      options: [
        'One',
        'Two',
        'Three',
        'Four',
        'Five',
        'Six',
        'Seven',
        'Eight',
        'Nine',
        'Ten',
        'Eleven',
        'Twelve',
      ],
      search: 't',
      title: 'Recent Searches',
    },
    overrides
  );
  const wrapper = shallow(<SearchSuggest {...props} />);

  return { wrapper, props };
};

describe('<SearchSuggest />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle className', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('search-suggest')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle isLoading', () => {
    const { wrapper } = setup({ isLoading: true });

    expect(wrapper.find('[data-test-id="search-suggest-loading"]')).toHaveLength(1);

    wrapper.setProps({ isLoading: false });

    expect(wrapper.find('[data-test-id="search-suggest-loading"]')).toHaveLength(0);
  });

  it('should handle isOpen', () => {
    const { wrapper } = setup({ isOpen: true });

    expect(wrapper.prop('isOpen')).toBe(true);

    wrapper.setProps({ isOpen: false });

    expect(wrapper.prop('isOpen')).toBe(false);
  });

  it('should handle limit', () => {
    const options = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l'];
    const { wrapper } = setup({ options, search: '' });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(10);

    wrapper.setProps({ limit: 5 });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(5);

    wrapper.setProps({ limit: null });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(10);
  });

  it('should handle onClear', () => {
    const { wrapper, props } = setup();

    wrapper.find('[data-test-id="search-suggest-clear"]').simulate('click');

    expect(props.onClear).toHaveBeenCalledTimes(1);
  });

  it('should handle onClick', () => {
    const { wrapper, props } = setup();

    wrapper.simulate('click');

    expect(props.onClick).toHaveBeenCalledTimes(1);
  });

  it('should handle onRemove', () => {
    const { wrapper, props } = setup();
    const stopPropagation = jest.fn();

    wrapper
      .find('[data-test-id="search-suggest-item"]')
      .at(0)
      .find('[data-test-id="search-suggest-remove"]')
      .simulate('click', { stopPropagation });
    wrapper
      .find('[data-test-id="search-suggest-item"]')
      .at(1)
      .find('[data-test-id="search-suggest-remove"]')
      .simulate('click', { stopPropagation });

    expect(stopPropagation).toHaveBeenCalledTimes(2);
    expect(props.onRemove).toHaveBeenCalledTimes(2);
    expect(props.onRemove).toHaveBeenNthCalledWith(1, 0);
    expect(props.onRemove).toHaveBeenNthCalledWith(2, 1);
  });

  it('should handle onSelect', () => {
    const options = ['A', 'B', 'C'];
    const { wrapper, props } = setup({ options, search: '' });

    wrapper
      .find('[data-test-id="search-suggest-item"]')
      .at(0)
      .simulate('click');

    wrapper
      .find('[data-test-id="search-suggest-item"]')
      .at(1)
      .simulate('click');

    expect(props.onSelect).toHaveBeenCalledTimes(2);
    expect(props.onSelect).toHaveBeenNthCalledWith(1, options[0], 0);
    expect(props.onSelect).toHaveBeenNthCalledWith(2, options[1], 1);
  });

  it('should handle options', () => {
    const options = ['A', 'b', 'c'];
    const { wrapper } = setup();

    wrapper.setProps({ options, search: '' });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(3);
  });

  it('should handle search', () => {
    const options = ['A', 'a', 'B', 'C'];
    const { wrapper } = setup({ options, search: 'a' });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(2);

    wrapper.setProps({ search: 'b' });

    expect(wrapper.find('[data-test-id="search-suggest-item"]')).toHaveLength(1);
  });

  it('should handle title', () => {
    const { wrapper } = setup({ title: '' });

    expect(wrapper.find('[data-test-id="search-suggest-header"]')).toHaveLength(0);

    wrapper.setProps({ title: 'Test' });

    expect(wrapper.find('[data-test-id="search-suggest-header"]')).toHaveLength(1);
  });

  it('should pass through other props', () => {
    const { wrapper } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });
});

describe('findMatches()', () => {
  it('should return true if option string includes search value', () => {
    expect(findMatches('Test String', 't')).toBe(true);
    expect(findMatches('Test String', 'T')).toBe(true);
    expect(findMatches('Test String', 'est')).toBe(true);
    expect(findMatches('Test String', 'ing')).toBe(true);
    expect(findMatches('Test String', 'Test S')).toBe(true);
  });

  it('should return false if option string does not include search value', () => {
    expect(findMatches('Test String', 'a')).toBe(false);
    expect(findMatches('Test String', 'z')).toBe(false);
    expect(findMatches('Test String', 'Tests')).toBe(false);
    expect(findMatches('Test String', 'TestString')).toBe(false);
  });
});

describe('highlightMatches()', () => {
  it('should return the original string if no search value', () => {
    expect(highlightMatches('foo')).toEqual('foo');
    expect(highlightMatches('foo', '')).toEqual('foo');
  });

  it('should wrap matching strings in <strong class="text-gray-900"> tags', () => {
    expect(highlightMatches('Test', 'T')).toEqual(
      '<strong class="text-gray-900">T</strong>es<strong class="text-gray-900">t</strong>'
    );
    expect(highlightMatches('Test', 't')).toEqual(
      '<strong class="text-gray-900">T</strong>es<strong class="text-gray-900">t</strong>'
    );
    expect(highlightMatches('Test', 'te')).toEqual('<strong class="text-gray-900">Te</strong>st');
    expect(highlightMatches('Test', 'est')).toEqual('T<strong class="text-gray-900">est</strong>');
  });
});
