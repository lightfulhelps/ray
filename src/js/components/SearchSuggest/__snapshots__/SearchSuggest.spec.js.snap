// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchSuggest /> should render 1`] = `
<DropdownMenu
  className="search-suggest"
  data-test-id="search-suggest"
  onClick={[MockFunction]}
>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-header"
    isHeader={true}
  >
    Recent Searches
     
    <div
      className="ml-1 text-danger cursor-pointer text-capitalize"
      data-test-id="search-suggest-clear"
      onClick={[MockFunction]}
    >
      Clear All
    </div>
  </DropdownItem>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-item"
    key="0"
    onClick={[Function]}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<strong>T</strong>wo",
        }
      }
    />
    <Icon
      className="ml-1"
      data-test-id="search-suggest-remove"
      name="close"
      onClick={[Function]}
    />
  </DropdownItem>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-item"
    key="1"
    onClick={[Function]}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<strong>T</strong>hree",
        }
      }
    />
    <Icon
      className="ml-1"
      data-test-id="search-suggest-remove"
      name="close"
      onClick={[Function]}
    />
  </DropdownItem>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-item"
    key="2"
    onClick={[Function]}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "Eigh<strong>t</strong>",
        }
      }
    />
    <Icon
      className="ml-1"
      data-test-id="search-suggest-remove"
      name="close"
      onClick={[Function]}
    />
  </DropdownItem>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-item"
    key="3"
    onClick={[Function]}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<strong>T</strong>en",
        }
      }
    />
    <Icon
      className="ml-1"
      data-test-id="search-suggest-remove"
      name="close"
      onClick={[Function]}
    />
  </DropdownItem>
  <DropdownItem
    className="d-flex justify-content-between align-items-center"
    data-test-id="search-suggest-item"
    key="4"
    onClick={[Function]}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<strong>T</strong>welve",
        }
      }
    />
    <Icon
      className="ml-1"
      data-test-id="search-suggest-remove"
      name="close"
      onClick={[Function]}
    />
  </DropdownItem>
</DropdownMenu>
`;
